
如何编译linux工程：
1. 纯手动编译:
	打开控制台，找到petalinux-tools的安装目录，在v2022.2中执行source settings.sh
	选择板子:
		export BOARD_SPECIFIC_DTS_FILE=radar_board_t4.dtsi  //或者radar_board_t5.dtsi 是T5板子
		export RADAR_BOARD=t4 //或者t5板子
		export BB_ENV_EXTRAWHITE="BOARD_SPECIFIC_DTS_FILE RADAR_BOARD"
	回到本目录下，执行petalinux-build，第一次编译时间会比较长
	编译完成后，会生成images文件夹，所有的image都在images/linux下
# source /opt/xilinx/petalinux-tools/settings.sh
# choose board:
	Cexport BOARD_SPECIFIC_DTS_FILE=radar_board_t4.dtsi  //radar_board_t5.dtsi 是T5板子
	export BB_ENV_EXTRAWHITE="BOARD_SPECIFIC_DTS_FILE"
# petalinux-build

2. 使用脚本编译
	./build_ci.sh all T5 或者 ./build_ci.sh all T4
	其中T5 和 T4是板子配置名称，默认是T5
	关于这个脚本的使用，运行./build_ci.sh help

如何更新xsa文件：
# petatlinux-config --get-hw-description ../rtos_app/tool/hw_spec 

如何生成qspi flash的烧录image
#sudo ./petalinux-xen-package-qspiflash.sh 
默认生成petalinux-xen-fw.bin，适用于64M的QSPI Flash, 可使用jtag直接烧录使用



编译问题处理：
1. a22804@a22804-G7-7700:~/petalinux-tools/v2022.2/zcu104/zcu104$ petalinux-build
[INFO] Sourcing buildtools
[INFO] Building project
[INFO] Sourcing build environment
[INFO] Generating workspace directory
INFO: bitbake petalinux-image-minimal
NOTE: Started PRServer with DBfile: /home/a22804/petalinux-tools/v2022.2/zcu104/zcu104/build/cache/prserv.sqlite3, Address: 127.0.0.1:43577, PID: 258647
Loading cache: 100% |###############################################################################################################################################################################| Time: 0:00:01
Loaded 6492 entries from dependency cache.
Parsing recipes: 100% |#############################################################################################################################################################################| Time: 0:00:01
Parsing of 4461 .bb files complete (4456 cached, 5 parsed). 6497 targets, 577 skipped, 1 masked, 0 errors.
NOTE: Resolving any missing task queue dependencies
Initialising tasks: 100% |##########################################################################################################################################################################| Time: 0:00:04
Checking sstate mirror object availability: 100% |##################################################################################################################################################| Time: 0:00:31
Sstate summary: Wanted 661 Local 18 Network 396 Missed 247 Current 1559 (62% match, 88% complete)
Removing 23 stale sstate objects for arch zynqmp_generic: 100% |####################################################################################################################################| Time: 0:00:00
NOTE: Executing Tasks
ERROR: u-boot-zynq-scr-1.0-r0 do_deploy: Error executing a python function in exec_func_python() autogenerated:

The stack trace of python calls that resulted in this exception/failure was:
File: 'exec_func_python() autogenerated', lineno: 2, function: <module>
     0001:
 *** 0002:plnx_deploy(d)
     0003:
File: '/home/a22804/petalinux-tools/v2022.2/zcu104/zcu104/components/yocto/layers/meta-petalinux/classes/plnx-deploy.bbclass', lineno: 215, function: plnx_deploy
     0211:    for package_bin in packageflags[pn].split():
     0212:        input, output = package_bin.split(':')
     0213:        inputfile = deploy_dir + '/' + input
     0214:        outputfile = output_path + '/' + output
 *** 0215:        copy_files(inputfile,outputfile)
     0216:
     0217:    symlinkpkgs = (d.getVar('SYMLINK_PACKAGES', True) or "").split()
     0218:    if pn in symlinkpkgs:
     0219:        for package in d.getVarFlag('SYMLINK_PACKAGES', pn).split():
File: '/home/a22804/petalinux-tools/v2022.2/zcu104/zcu104/components/yocto/layers/meta-petalinux/classes/plnx-deploy.bbclass', lineno: 170, function: copy_files
     0166:       if not os.path.exists(os.path.dirname(outputfile)):
     0167:           os.mkdir(os.path.dirname(outputfile))
     0168:       if os.path.isdir(inputfile):
     0169:           if os.path.exists(outputfile):
 *** 0170:               shutil.rmtree(outputfile)
     0171:           shutil.copytree(inputfile, outputfile,symlinks=True)
     0172:       else:
     0173:           shutil.copy2(inputfile,outputfile)
     0174:
File: '/home/a22804/petalinux-tools/v2022.2/components/yocto/buildtools/sysroots/x86_64-petalinux-linux/usr/lib/python3.9/shutil.py', lineno: 737, function: rmtree
     0733:                try:
     0734:                    # symlinks to directories are forbidden, see bug #1669
     0735:                    raise OSError("Cannot call rmtree on a symbolic link")
     0736:                except OSError:
 *** 0737:                    onerror(os.path.islink, path, sys.exc_info())
     0738:        finally:
     0739:            os.close(fd)
     0740:    else:
     0741:        try:
File: '/home/a22804/petalinux-tools/v2022.2/components/yocto/buildtools/sysroots/x86_64-petalinux-linux/usr/lib/python3.9/shutil.py', lineno: 735, function: rmtree
     0731:                    onerror(os.rmdir, path, sys.exc_info())
     0732:            else:
     0733:                try:
     0734:                    # symlinks to directories are forbidden, see bug #1669
 *** 0735:                    raise OSError("Cannot call rmtree on a symbolic link")
     0736:                except OSError:
     0737:                    onerror(os.path.islink, path, sys.exc_info())
     0738:        finally:
     0739:            os.close(fd)
Exception: OSError: Cannot call rmtree on a symbolic link

ERROR: Logfile of failure stored in: /home/a22804/petalinux-tools/v2022.2/zcu104/zcu104/build/tmp/work/zynqmp_generic-xilinx-linux/u-boot-zynq-scr/1.0-r0/temp/log.do_deploy.259152
ERROR: Task (/home/a22804/petalinux-tools/v2022.2/zcu104/zcu104/components/yocto/layers/meta-xilinx/meta-xilinx-core/recipes-bsp/u-boot/u-boot-zynq-scr.bb:do_deploy) failed with exit code '1'
NOTE: Tasks Summary: Attempted 5624 tasks of which 5613 didn't need to be rerun and 1 failed.

Summary: 1 task failed:
  /home/a22804/petalinux-tools/v2022.2/zcu104/zcu104/components/yocto/layers/meta-xilinx/meta-xilinx-core/recipes-bsp/u-boot/u-boot-zynq-scr.bb:do_deploy
Summary: There was 1 ERROR message shown, returning a non-zero exit code.
ERROR: Failed to build project. Check the /home/a22804/petalinux-tools/v2022.2/zcu104/zcu104/build/build.log file for more details...
a22804@a22804-G7-7700:~/petalinux-tools/v2022.2/zcu104/zcu104$ 

A: 删除文件夹images就可以了。是pythone 2.7导致的问题
